OutOfSchool/OutOfSchool.WebApi/Services/ProviderService.cs

```diff
@@ -108,6 +115,57 @@ public async Task<ProviderDto> Create(ProviderDto providerDto)
             return newProvider.ToModel();
         }
 
+        /// <inheritdoc/>
+        public async Task<ProviderDto> CreateV2(ProviderDto providerDto)
+        {
+            _ = providerDto ?? throw new ArgumentNullException(nameof(providerDto));
+
+            logger.LogDebug("Provider creating was started");
+
+            if (providerRepository.ExistsUserId(providerDto.UserId))
+            {
+                throw new InvalidOperationException(localizer["You can not create more than one account."]);
+            }
+
+            // Note: clear the actual address if it is equal to the legal to avoid data duplication in the database
+            if (providerDto.LegalAddress.Equals(providerDto.ActualAddress))
+            {
+                providerDto.ActualAddress = null;
+            }
+
+            var providerDomainModel = providerDto.ToDomain();
+
+            // BUG: concurrency issue:
+            //      while first repository with this particular user id is not saved to DB - we can create any number of repositories for this user.
+            if (providerRepository.SameExists(providerDomainModel))
+            {
+                throw new InvalidOperationException(localizer["There is already a provider with such a data"]);
+            }
+
+            var users = await usersRepository.GetByFilter(u => u.Id.Equals(providerDto.UserId)).ConfigureAwait(false);
```

Comment:
All code before `var users = await usersRepository.GetByFilter...` seems to be related to `providerDto` validation and copy of same code in `Create` method. Could you extract method and thus avoid code duplication?